src/actions/index.js

```diff
@@ -163,15 +184,178 @@ export const devicesDelete = (deviceId) => {
+      })
+      .then(entity => {
+        dispatch(action('DEVICES_CREATE', newDevice));
+        dispatch(loading(false));
+      })
+      .catch(err => {
+        console.log('entity seems to be there already...');
+        dispatch(loading(false));
+      });
+  };
+}
+
+// fetch all curent user data
+export const fetchCurrentUser = () => {
+  return (dispatch) => {
+    dispatch(loading(true))
+    agile.idm.user.getCurrentUserInfo()
+      .then(user => {
+        dispatch(action('CURRENT_USER', user));
+        dispatch(loading(false));
+      })
+      .catch(err => {
+        errorHandle(err, dispatch)
+      });
+  };
+}
+
+export const userFetch = (id) => {
+  return (dispatch) => {
+    dispatch(loading(true))
+    agile.idm.entity.get(id, 'user')
+      .then(user => {
+        dispatch(action('USER', user));
+        dispatch(loading(false));
+      })
+      .catch(err => {
+        errorHandle(err, dispatch)
+      });
+  };
+}
+
+export const setPassword = (params) => {
+  return (dispatch) => {
+    dispatch(loading(true))
+    agile.idm.entity.setAttribute(params) // TODO set password through user in stack
+      .then(entity => {
+        dispatch(action('ENTITY_ATTRIBUTE_SET', entity));
+        dispatch(loading(false));
+      })
+      .catch(err => {
+        errorHandle(err, dispatch)
+      });
+  };
+}
+
+export const setEntityData = (params) => {
+  return (dispatch) => {
+    dispatch(loading(true))
+    agile.idm.entity.setAttribute(params)
+      .then(entity => {
+        dispatch(action('ENTITY_ATTRIBUTE_SET', entity));
+        dispatch(loading(false));
+      })
+      .catch(err => {
+        errorHandle(err, dispatch)
+      });
+  };
+}
+
+export const deleteAttribute = (params) => {
+  return (dispatch) => {
+    dispatch(loading(true))
+    agile.idm.entity.deleteAttribute(params.id, params.type, params.attribute)
+      .then(entity => {
+        dispatch(action('ENTITY_ATTRIBUTE_SET', entity));
+        dispatch(loading(false));
+      })
+      .catch(err => {
+        errorHandle(err, dispatch)
+      });
+  };
+}
+
+export const canExecuteActions = (id, type, attribute_names, actions) => {
+  return (dispatch) => {
+    dispatch(loading(true))
+    var queryObject = [];
+    var actions_seen = [];
+    for (var act in actions) {
+      if (actions.hasOwnProperty(act) && actions_seen.indexOf(act) === -1) {
+        actions_seen.push(act);
+      }
+    }
+    attribute_names.push('actions.self'); //Add root object
+    attribute_names.forEach(attribute => {
+      actions_seen.forEach(method => {
+        queryObject.push({entityId: id, entityType: type.startsWith('/') ? type : '/' + type, method: method, field: attribute});
+      });
     });
+
+    agile.policies.pdp.evaluate(queryObject).then(
+      results => {
+        var result_list = {id: id, type: type, policies: {}};
+        var i = 0;
+        attribute_names.forEach(attribute => {
```

Comment:
Not exactly sure what these loops are doing, but perhaps check if there is a lodash function to simplify the logic?

Reply:
This is an optimization that we had to do, to query less times the policies. Otherwise we end up querying the same thing too many times. This could have performance impact on the PDP, considering that it only uses a levelDB database. So, I am afraid there is no way around this, if we want to keep an eye on resource consumption on the gateway's side.

Reply:
This is an optimization that we had to do, to query less times the policies. Otherwise we end up querying the same thing too many times. This could have performance impact on the PDP, considering that it only uses a levelDB database. So, I am afraid there is no way around this, if we want to keep an eye on resource consumption on the gateway's side.